drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,1597 @@
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief check for MMC card presence
+ *
+ * Checks if card is present on the bus.
+ *
+ * @param dev MMC device
+ *
+ * @retval 1 card is present
+ * @retval 0 card is not present
+ * @retval -EINVAL: some of pointers provided to the function are NULL
+ */
+static int rcar_mmc_get_card_present(const struct device *dev)
+{
+	const struct mmc_rcar_cfg *cfg;
+
+	if (!dev || !dev->config) {
+		return -EINVAL;
+	}
+
+	cfg = dev->config;
+	if (cfg->non_removable) {
+		return 1;
+	}
+
+	return !!(rcar_mmc_read_reg32(cfg, RCAR_MMC_INFO1) & RCAR_MMC_INFO1_CD);
+}
+
+/**
+ * @brief run MMC tuning
+ *
+ * MMC cards require signal tuning for UHS modes SDR104, HS200 or HS400.
+ * This function allows an application to request the SD host controller
+ * to tune the card.
+ *
+ * @param dev MMC device
+ *
+ * @retval 0 tuning succeeded, card is ready for commands
+ * @retval -ENOTSUP: controller does not support tuning
+ * @retval -EINVAL: the dev pointer is NULL
+ */
+static int rcar_mmc_execute_tuning(const struct device *dev)
+{
+	if (!dev) {
+		return -EINVAL;
+	}
+
+	LOG_ERR("%s: tuning isn't supported yet", dev->name);
+	return -ENOTSUP;
+}
+
+/**
+ * @brief Get MMC controller properties
+ *
+ * Gets host properties from the host controller. Host controller should
+ * initialize all values in the @ref sdhc_host_props structure provided.
+ *
+ * @param dev Renesas MMC device
+ * @param props property structure to be filled by MMC driver
+ *
+ * @retval 0 function succeeded.
+ * @retval -EINVAL: some of pointers provided to the function are NULL
+ */
+static int rcar_mmc_get_host_props(const struct device *dev,
+	struct sdhc_host_props *props)
+{
+	struct mmc_rcar_data *data;
+
+	if (!props || !dev || !dev->data) {
+		return -EINVAL;
+	}
+
+	data = dev->data;
+	memcpy(props, &data->props, sizeof(*props));
+	return 0;
+}
+
+static const struct sdhc_driver_api rcar_sdhc_api = {
+	.card_busy = rcar_mmc_card_busy,
+	.execute_tuning = rcar_mmc_execute_tuning,
+	.get_card_present = rcar_mmc_get_card_present,
+	.get_host_props = rcar_mmc_get_host_props,
+	.request = rcar_mmc_request,
+	.reset = rcar_mmc_reset,
+	.set_io = rcar_mmc_set_io,
+};
+
+/*
+ * TODO: share structure using header file or add API for get base
+ *       address to CPG driver
+ */
+struct r8a7795_cpg_mssr_config {
+	mm_reg_t base_address;
+};
+
+/**
+ * @brief Start SD-IF2 clock at 200MHz
+ *
+ * @param cfg The Renesas MMC driver configuration
+ *
+ * @retval 0 on success, negative errno code otherwise
+ */
+static int rcar_mmc_init_start_clk(const struct mmc_rcar_cfg *cfg)
+{
+	unsigned int key;
+	int ret = 0;
+	const struct device *cpg_dev = cfg->cpg_dev;
+	const struct r8a7795_cpg_mssr_config *clk_cfg = cpg_dev->config;
+
+	key = irq_lock();
+
+	/*
+	 * TODO: fix me for Gen4
+	 *
+	 * RCAR Gen3 SD-IF2 clock frequency control register base + SDIF2_CFCR_OFFSET
+	 * SDn = 200 MHz
+	 */
+	sys_write32(1, clk_cfg->base_address + SDIF2_CFCR_OFFSET);
```

Comment:
What makes this `sys_write32` so different that you need to disable interrupts there?

Reply:
I have used the same approach as in CPG driver

Reply:
in cpg driver, it is necessary for ensuring the atomicity of several consecutive writes/reads to/from a cpg register, but we don't need irq disabling here, because we have only one write operation in a row